CheckedPtr should use zombie mode

geoffreygaren opened the pull request:
#### bf47ad4f9e351004a197f61b30dae13c11a03c37
<pre>
CheckedPtr should use zombie mode
<a href="https://github.com/WebKit/WebKit/pull/27212">https://github.com/WebKit/WebKit/pull/27212</a>

Reviewed by Ryosuke Niwa.

Previously a dangling CheckedPtr triggered a crash in the destructor of the
pointed-to object. That kind of crash was hard to debug, since the backtrace
didn&apos;t report any information about the pointer. As a result, we&apos;ve had to limit
CheckedPtr adoption.

Now a dangling CheckedPtr triggers a crash in the CheckedPtr&apos;s destructor,
so the backtrace points the finger at the dangling pointer. Now, we can use
CheckedPtr as much as we like.

Note: We explicitly check-and-crash in the CheckedPtr&apos;s destructor, but in
release builds we don&apos;t explicitly check-and-crash in the CheckedPtr&apos;s
accessor methods. This maintains CheckedPtr&apos;s feature of having nearly
identical performance to a raw pointer.

Even though we don&apos;t explicitly check-and-crash in the CheckedPtr&apos;s
accessor methods, all accesses are memory safe. The implementation
detects a dangling CheckedPtr, and then zeroes and leaks the pointed-to
object instead of freeing it. Therefore, any attempt to dereference a
pointer data member in the pointed-to object deterministically crashes
with a nullptr dereference.

This change requires putting a special macro in every subclass in a class
hierarchy that uses CheckedPtr, since we need to override operator delete
in order to leak the pointed-to object. (If you forget the macro entirely,
you&apos;ll get a compile-time failure. If you remember the macro in a base
class but forget it in a subclass, you&apos;ll get a debug assertion. This is
similar to how IsoHeaps work.)

When we check ptrCount() at destruction time, we use a new function,
SingleThreadIntegralWrapper::valueWithoutThreadCheck(). We use this
because checking happens after ~SingleThreadIntegralWrapper() runs, which
means that SingleThreadIntegralWrapper::m_thread is null. It&apos;s thread-safe
to stop doing thread checks after destruction; in theory no thread has any
outstanding reference anymore.

Canonical link: <a href="https://commits.webkit.org/277590@main">https://commits.webkit.org/277590@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/13fea38c5c06a2be40b3c9b10d89d52c425ce281

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48043 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27248 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/50971 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50730 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44104 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/50350 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/33136 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/24742 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/39278 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48626 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/24918 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/41557 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20388 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22399 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/42733 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6097 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/41340 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/44388 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/43168 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52626 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/47537 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/23091 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/19450 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/46602 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24356 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/41710 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45455 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10604 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25156 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/55032 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/24078 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11316 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WTF/wtf/CheckedRef.h:
```diff
@@ -323,14 +323,28 @@ template <typename StorageType, typename PtrCounterType> class CanMakeCheckedPtr
+        // Zero is a convenient sentinel because:
+        //     * it can't appear in normal execution;
+        //     * it triggers a crash at the point where the innermost dangling CheckedPtr goes out of scope.
+        RELEASE_ASSERT(uncheckedPtrCount());
```
Don't we also want to check this when CheckedPtr is accessed in `ptr()`?

geoffreygaren replied:
It's a tradeoff. I'm open to opinions / preferences here.

The argument against checking inside ptr() is:
* it's more efficient not to check, and the current CheckedPtr doesn't check
* not checking is still memory safe because we've zeroed the object, so there's no way to load a bad pointer out of it -- if you load and dereference a pointer, you'll just get a nullptr crash

(For context, Chrome has stopped paying bug bounties for UAF of zeroed+leaked objects.)

The argument for checking inside ptr() is:
* if you load a non-pointer value, you crash sooner, which can sometimes mean a clearer backtrace / bug symptom
* if you load and dereference a pointer value, you crash with a clearer assertion failure instead of a more mysterious nullptr

My understanding is that @smfr and @alanbaradlay prefer not checking because it's more efficient, so I wanted to preserve that feature for them. I wouldn't want someone to feel like they needed to create a FastCheckedPtr or something like that.

But I notice now that I can definitely add a debug ASSERT in ptr(). So I'll at least do that.

rniwa replied:
Thanks for the clarification. It seems that we should clarify this in the commit message.

geoffreygaren replied:
Sounds good. I'll update it.

cdumez commented on Source/WebKitLegacy/WebCoreSupport/WebResourceLoadScheduler.h:
```diff
@@ -83,8 +83,10 @@ class WebResourceLoadScheduler final : public WebCore::LoaderStrategy {
-        WTF_MAKE_NONCOPYABLE(HostInformation); WTF_MAKE_FAST_ALLOCATED;
+    class HostInformation final : public CanMakeWeakPtr<HostInformation>, public CanMakeCheckedPtr<HostInformation> {
+        WTF_MAKE_NONCOPYABLE(HostInformation);
+        WTF_MAKE_FAST_ALLOCATED;
```
Shouldn't we have `WTF_MAKE_FAST_ALLOCATED` on CanMakeCheckedPtr instead so we don't have to specify it for every subclass?

geoffreygaren replied:
Some classes use WTF_MAKE_ISO_ALLOCATED or WTF_MAKE_FAST_ALLOCATED_WITH_HEAP_IDENTIFIER.

Other subclasses have an incompatible definition of memory allocation. For example, GlyphBuffer is stack-allocated; and WebProcessPool uses API::ObjectImpl's Objective-C allocator. 

I was hoping that putting the allocation specification in each subclass would help clarify the requirements. 

I see now that RefCounted specifies WTF_MAKE_FAST_ALLOCATED though. I guess it provides a good default that way. But CanMakeWeakPtr doesn't. 

I guess I'm not sure. Should I add it?

cdumez commented on Source/WebKitLegacy/WebCoreSupport/WebResourceLoadScheduler.h:
```diff
@@ -83,8 +83,10 @@ class WebResourceLoadScheduler final : public WebCore::LoaderStrategy {
 
-    class HostInformation : public CanMakeWeakPtr<HostInformation>, public CanMakeCheckedPtr<HostInformation> {
-        WTF_MAKE_NONCOPYABLE(HostInformation); WTF_MAKE_FAST_ALLOCATED;
+    class HostInformation final : public CanMakeWeakPtr<HostInformation>, public CanMakeCheckedPtr<HostInformation> {
```
Shouldn't this be `CanMakeCheckedPtr` instead of `CanMakeCheckedPtr<HostInformation>`? Same comment may apply elsewhere.

geoffreygaren replied:
CanMakeCheckedPtr requires the subclass type as a template parameter in order to verify that the type includes the WTF_OVERRIDE_DELETE_FOR_CHECKED_PTR macro. 

Is there some fancy C++ I need to learn? How would "CanMakeCheckedPtr" infer the type parameter?

rniwa commented on Source/WebCore/Modules/fetch/FetchLoader.h:
```diff
@@ -43,6 +43,8 @@ class ScriptExecutionContext;
 
 class WEBCORE_EXPORT FetchLoader final : public ThreadableLoaderClient {
+    WTF_MAKE_FAST_ALLOCATED_WITH_HEAP_IDENTIFIER(Loader);
+    WTF_OVERRIDE_DELETE_FOR_CHECKED_PTR(FetchLoader);
```
What happens if someone creates a subclass and forget to use these macros?
Do we get a compilation error/warning? Or will it silently fail?

geoffreygaren replied:
If a class hierarchy forgets WTF_OVERRIDE_DELETE_FOR_CHECKED_PTR entirely, you get a compilation error.

If a base class includes WTF_OVERRIDE_DELETE_FOR_CHECKED_PTR, but a subclass forgets it, then you get a runtime assertion when you delete the subclass object. (I tried hard to make that a compilation error, but I couldn't figure a way. Ultimately I decided that a runtime assertion is OK, since you get a deterministic failure in any debug run of the regression tests. IsoHeaps behave similarly.)

rniwa commented on Source/WTF/wtf/CheckedRef.h:
```diff
@@ -323,14 +336,26 @@ template <typename StorageType, typename PtrCounterType> class CanMakeCheckedPtr
+        --m_count;
+    }
+
+    ALWAYS_INLINE PtrCounterType uncheckedPtrCount() const
```
Hm... what does "unchecked" mean in this context? Isn't it more straight forward to call it `checkedPtrCount`?

geoffreygaren replied:
"Unchecked" means that the single-threaded CanMakeCheckedPtr will skip its "same thread" assertion. We need this because we call uncheckedPtrCount() after the destructor has run, when SingleThreadIntegralWrapper::m_thread is null.

rniwa replied:
I see. Perhaps clarification in the commit message or a comment will help clarify.

rniwa commented on Source/WTF/wtf/SingleThreadIntegralWrapper.h:
```diff
@@ -40,6 +40,8 @@ class SingleThreadIntegralWrapper {
     SingleThreadIntegralWrapper& operator++();
     SingleThreadIntegralWrapper& operator--();
 
+    IntegralType uncheckedGet() const { return m_value; }
```
It's rather confusing to call this "uncheckedGet". What is unchecked in what sense isn't clear specially when it's used with CheckedPtr. But why do we need this variant to begin with?
Why can we not check the thread identity?

geoffreygaren replied:
I kind of agree that this is confusing. Not sure how to improve it.

I could change ptrCount() to have a function argument or template parameter `enum class DoThreadCheck { No, Yes }`. Or I could rename uncheckedPtrCount() to ptrCountWithoutThreadCheck(). Does one of those options seem better to you?

rniwa replied:
Renaming it to `ptrCountWithoutThreadCheck` sounds good to me.

geoffreygaren committed:
CheckedPtr should use zombie mode
https://github.com/WebKit/WebKit/pull/27212

Reviewed by Ryosuke Niwa.

Previously a dangling CheckedPtr triggered a crash in the destructor of the
pointed-to object. That kind of crash was hard to debug, since the backtrace
didn't report any information about the pointer. As a result, we've had to limit
CheckedPtr adoption.

Now a dangling CheckedPtr triggers a crash in the CheckedPtr's destructor,
so the backtrace points the finger at the dangling pointer. Now, we can use
CheckedPtr as much as we like.

Note: We explicitly check-and-crash in the CheckedPtr's destructor, but in
release builds we don't explicitly check-and-crash in the CheckedPtr's
accessor methods. This maintains CheckedPtr's feature of having nearly
identical performance to a raw pointer.

Even though we don't explicitly check-and-crash in the CheckedPtr's
accessor methods, all accesses are memory safe. The implementation
detects a dangling CheckedPtr, and then zeroes and leaks the pointed-to
object instead of freeing it. Therefore, any attempt to dereference a
pointer data member in the pointed-to object deterministically crashes
with a nullptr dereference.

This change requires putting a special macro in every subclass in a class
hierarchy that uses CheckedPtr, since we need to override operator delete
in order to leak the pointed-to object. (If you forget the macro entirely,
you'll get a compile-time failure. If you remember the macro in a base
class but forget it in a subclass, you'll get a debug assertion. This is
similar to how IsoHeaps work.)

When we check ptrCount() at destruction time, we use a new function,
SingleThreadIntegralWrapper::valueWithoutThreadCheck(). We use this
because checking happens after ~SingleThreadIntegralWrapper() runs, which
means that SingleThreadIntegralWrapper::m_thread is null. It's thread-safe
to stop doing thread checks after destruction; in theory no thread has any
outstanding reference anymore.

Canonical link: https://commits.webkit.org/277590@main
